forem/forem
# [Search 2.0] Optimize search reading list

## What type of PR is this? (check all applicable)


## Description

This PR is the result of conversations we had in private and the various experiments made in https://github.com/forem/forem/pull/13193

To summarize: we decided to write a trigger that would build the search document for each article. This trigger runs "before insert or update" and collects the following fields (maintaining feature parity with the Elasticsearch based UX): `articles.body_markdown`, `articles.cached_tag_list`, `articles.title`, the `name` in `articles.cached_organization`, `articles.cached_user_name` and `articles.cached_user_username`.

Initially the trigger was joining `organizations.name` and `users.name` + `users.username` but while writing specs I discovered the cached fields in `articles`. This greatly simplified the trigger :-) 

Since triggers are per table, we rely on two assumptions in the business logic:

- for `articles.cached_organization` which comes from the `articles <-> organizations` relation, we rely on a callback that fires whenever `organizations.name` is changed that will update the org's articles
- for the user fields which come from the `articles <-> users` relation, we rely on `Profiles::Update.call` that will resave articles when `users.name` or `users.username` are updated

The drawback to this trigger based approach (thanks to [this answer](https://dba.stackexchange.com/a/289362/226575)) on DBA StackExchange is the following:

> if you restore the database from a dump, the tables won't be restores in the order you expect, and you could end up with values missing

We'd need to resave all articles if we do a table restore in an emergency situation. I don't think it's enough case to block this feature, but we should take this into account and add this note to our recovery procedures.

## Related Tickets & Documents

- https://github.com/forem/forem/pull/13193
- https://dba.stackexchange.com/questions/289356/trigger-only-fires-50-of-the-time/289362

## QA Instructions, Screenshots, Recordings

1. run the migration and run `bin/setup` to execute the data update script
2. add articles to your local reading list
3. go to http://localhost:3000/readinglist and use the page: archive items, unarchive them, filter by multiple tags, search for content with and without tags selected
4. add an article you wrote
5. go to http://localhost:3000/settings and change your name or username
6. go back to the reading list, the new name should be reflected in your article, make sure you can use it as a search keyword
7. create an organization
8. publish an article under that organization
9. add the article to the reading list and find it by inputing the name of the org in the search box
9. go to http://localhost:3000/settings/organization and change the organization's name
10. go back to the reading list and enter the new name in th search box

### UI accessibility concerns?

Nope

## Added tests?

      have not been included_

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
      or in a [forem.dev](http://forem.dev) post
      replace this line with details on why this change doesn't need to be
      shared_
